File: Scala/src/main/scala/factory/simple/SimplePizzaFactory.scala
package factory.simple

import factory.PizzaType
import factory.PizzaType.PizzaType

trait SimplePizzaFactory {

  def createPizza(pizzaType: PizzaType): Pizza

}

class NyPizzaFactory extends SimplePizzaFactory {
  override def createPizza(pizzaType: PizzaType): Pizza = pizzaType match {
    case PizzaType.CHEESE => new NyCheesePizza()
    case PizzaType.PEPPERONI => new NyPepperoniPizza()
    case PizzaType.CLAM => new NyClamPizza()
    case PizzaType.VEGGIE => new NyVeggiePizza()
  }
}

class ChicagoPizzaFactory extends SimplePizzaFactory {
  override def createPizza(pizzaType: PizzaType): Pizza = pizzaType match {
    case PizzaType.CHEESE => new ChicagoCheesePizza()
    case PizzaType.PEPPERONI => new ChicagoPepperoniPizza()
    case PizzaType.CLAM => new ChicagoClamPizza()
    case PizzaType.VEGGIE => new ChicagoVeggiePizza()
  }
}

File: Kotlin/src/main/kotlin/chapter4_factory/simple_factory/SimplePizzaFactory.kt
package chapter4_factory.simple_factory

class SimplePizzaFactory {
    fun createPizza(type: String): Pizza {
        return when(type) {
            "cheese" -> CheesePizza()
            "pepperoni" -> PepperoniPizza()
            "clam" -> ClamPizza()
            "veggie" -> VeggiePizza()
            else -> throw IllegalStateException()
        }
    }
}